lib/discover_endpoint.js

```diff
@@ -0,0 +1,340 @@
+var AWS = require('./core');
+var util = require('./util');
+var ENDPOINT_OPERATION_MAX_RETRIES = 60;
+var endpointDiscoveryEnabledEnv = 'AWS_ENABLE_ENDPOINT_DISCOVERY';
+
+/**
+ * Generate key to index the endpoints in the cache
+ * @return [map<String,String>] object with keys to index endpoints.
+ * @api private
+ */
+function getCacheKey(request) {
+  var service = request.service;
+  var api = service.api || {};
+  var operations = api.operations;
+  var identifiers = {};
+  if (operations[request.operation] && operations[request.operation].name) {
+    identifiers.operation = operations[request.operation].name;
+  }
+  if (service.config.region) {
+    identifiers.region = service.config.region;
+  }
+  if (api.serviceId) {
+    identifiers.serviceId = api.serviceId
+  }
+  if (service.config.credentials.accessKeyId) {
+    identifiers.accessKeyId = service.config.credentials.accessKeyId
+  }
+  return identifiers;
+}
+
+/**
+ * @api private
```

Comment:
Can you add a comment that explains what this function is supposed to do? It looks like you're populating an object with identifiers and customer-provided values, but it took me a while to grok that and the function name isn't clear.